### PR TITLE
[PM-4270] Individual cipher key encryption

### PR DESCRIPTION
**Pull request by dani-garcia**
## Type of change
```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective
Implement individual vault item encryption and encryption. Note that at the moment it will not migrate items that don't have a key when encrypting.

### CODE REVIEW

File: crates/bitwarden/src/vault/cipher/cipher.rs [hunk 1]
```diff
@@ -43,6 +43,8 @@ pub struct Cipher {
     pub folder_id: Option<Uuid>,
     pub collection_ids: Vec<Uuid>,
 
+    pub key: Option<EncString>,
```
Comment (Hinton) [inline crates/bitwarden/src/vault/cipher/cipher.rs:46]: Can we add a comment here what the key is.

```suggestion
    /// More recent ciphers uses individual encryption keys to encrypt the other fields of the Cipher.
    pub key: Option<EncString>,
```

File: crates/bitwarden/src/vault/cipher/cipher.rs [hunk 2]
```diff
@@ -77,6 +79,8 @@ pub struct CipherView {
...
```
Comment (Hinton) [inline crates/bitwarden/src/vault/cipher/cipher.rs:82]: Do we want to expose the key in a view?

File: crates/bitwarden/src/vault/cipher/cipher.rs [hunk 3]
```diff
@@ -131,11 +135,15 @@ pub struct CipherListView {
 
 impl KeyEncryptable<Cipher> for CipherView {
     fn encrypt_with_key(self, key: &SymmetricCryptoKey) -> Result<Cipher> {
+        let key_owned = Cipher::get_cipher_key(key, &self.key)?;
+        let key = key_owned.as_ref().unwrap_or(key);
```
Comment (Hinton) [inline crates/bitwarden/src/vault/cipher/cipher.rs:139]: **suggestion**: I think we can improve the clarify here with some better variable names.

```suggestion
        let ciphers_key = Cipher::get_cipher_key(key, &self.key)?;
        let key = ciphers_key.as_ref().unwrap_or(key);
```

File: crates/bitwarden/src/vault/cipher/cipher.rs [hunk 4]
```diff
@@ -190,6 +202,19 @@ impl KeyDecryptable<CipherView> for Cipher {
...
+        inner_key: &Option<EncString>,
+    ) -> Result<Option<SymmetricCryptoKey>> {
+        inner_key
+            .as_ref()
+            .map(|k| {
+                let key: Vec<u8> = k.decrypt_with_key(key)?;
+                SymmetricCryptoKey::try_from(key.as_slice())
+            })
+            .transpose()
+    }
```
Comment (Hinton) [inline crates/bitwarden/src/vault/cipher/cipher.rs:216]: **suggestion:** Can we add a comment what this function is doing?

File: crates/bitwarden/src/vault/cipher/cipher.rs (hunk 2 above)
Comment (dani-garcia) [reply crates/bitwarden/src/vault/cipher/cipher.rs:82]: If we want to encrypt it back with the same Key, yeah. Though I guess we can generate a new key each time we encrypt a CipherView into a Cipher, if we wanted.

Comment (dani-garcia) [reply crates/bitwarden/src/vault/cipher/cipher.rs:83]: Should we add a TODO or open a ticket for the future?

**Review by Hinton — APPROVED** [review]

**Pull request merged**